### PR TITLE
Add Specializations for SortedRange

### PR DESCRIPTION
**Pull request by nordlow**
Add new predicate function `std.range.primitives.isSortedRange`: **DONE**

Add specializations for `SortedRange` in the followings algorithms:
- `sort`: just return input argument if `pred` parameter matches: **DONE**
- `sort`-alikes: just return input argument if `pred` parameter matches:
- `isSorted`: return `true`: **DONE**
- `find` and alikes, should do some kind of binary search. These could take an extra param `SearchPolicy` for sorted ranges that defaults to binary search.
- Some of the functions that return a sub-range or a mutated range could probably be returning sorted ranges as well if its input is a sorted range, `remove`, `strip` and `split` at least could.
- `minPos`: return input as is, if `pred` matches: **DONE**

In compliance with C++ STL also add
- `minElement`: O(1) for `SortedRange`, O(n) otherwise: **DONE**
- `maxElement`: O(1) for `SortedRange!BidirectionalRange`, O(n) otherwise: **DONE**
- `minmaxElement`: O(1) for `SortedRange!BidirectionalRange`, O(n) otherwise.

See also: http://forum.dlang.org/post/yenezfjjteokzyvgmzcf@forum.dlang.org
See also: https://issues.dlang.org/show_bug.cgi?id=11667
See also: http://forum.dlang.org/post/mqskao$28vh$1@digitalmars.com

See also: http://en.cppreference.com/w/cpp/algorithm/min_element

### CODE REVIEW

**Comment by DmitryOlshansky** [comment]
> O(2)

Hilarious. It's sill O(1), the number just means it's `some constant factor times 1` so it doesn't matter if this number is 1, 2, 1000 since multiplier in front of it is unspecified.

**Comment by DmitryOlshansky** [comment]
> std.range.primitives.isSortedRange

I recall what the problem we had with it was - literally things like `SortedRange!(int[], binaryFun!" a < b")` and `SortedRange!(int[], (a,b) => a < b)` and `SortedRange!(int[], (b,c) => b < c)` are distinct types b/c respective predicates are not identical.

It doesn't block anything but should be taken into consideration, a compiler ought to try and match lambas with identical bodies at least token-wise identical.

**Comment by nordlow** [comment]
> Hilarious. It's sill O(1)

I kind of the new that ;)

**Comment by nordlow** [comment]
Should `minElement`, `maxElement` and `minmaxElement` return a range instead?

File: std/algorithm/searching.d [hunk 1]
```diff
@@ -2777,23 +2777,30 @@ smallest element and with the same ending as $(D range). The function
 can actually be used for finding the maximum or any other ordering
 predicate (that's why $(D maxPos) is not provided).
  */
-Range minPos(alias pred = "a < b", Range)(Range range)
+auto minPos(alias pred = "a < b", Range)(Range range)
```
Comment (JakobOvrum) [inline std/algorithm/searching.d]: The return type is `Range` for all paths; explicit return type is better for documented functions as it provides important information to readers.

File: std/algorithm/searching.d [hunk 2]
```diff
@@ -2852,6 +2864,76 @@ unittest
     assert(minPos!("a[0] < b[0]")(b) == [ [2, 4], [4], [4] ]);
 }
 
+import std.algorithm.comparison : min, max;
+
+/** Returns: Minimum Element in $(D range) or $(D unit) if $(D range) is empty.
+ */
+auto minElement(alias F = min, R)(R range,
```
Comment (JakobOvrum) [inline std/algorithm/searching.d]: Do we really need these? "Compliance with C++ STL" doesn't seem like a good argument. I recommend moving these to a different PR where they can be evaluated separately.

**Comment by JakobOvrum** [comment]
Functions that return subranges often already them as the same type as the whole range; this already works with `SortedRange`.

Comparing predicates has been discussed a lot before, particularly in the context of string lambdas. One such thread:
http://forum.dlang.org/post/jnlqesrwxfekdsxjerlp@forum.dlang.org (Sorry, I have a hard time digging up the older ones)

File: std/algorithm/searching.d (hunk 1 above)
Comment (nordlow) [inline std/algorithm/searching.d]: Done.

**Comment by nordlow** [comment]
@JakobOvrum I guess you're referring to `remove`, `strip` and `split`, right? Should we add some static asserts to those functions to verify that `SortedRange` is propagated correctly?

**Comment by JackStouffer** [comment]
@nordlow if you ever solve the comparing predicates problem, `isPermutation` once had a `SortedRange` overload that could be added back in to make the operation `O(1)` by basically acting as an wrapper to `equal`, but it was dropped because the function could not determine if the two predicates for the `SortedRange`s were equal, and was therefore buggy.

**Comment by JakobOvrum** [comment]
@nordlow, algorithms that return the input range type as-is in some capacity, of which there are quite a few, tend to guarantee this in the documentation (see `find`, `strip`, and even composite return types like `findSplit` and `retro`). This is an integral part of their interface, orthogonal to `SortedRange` in particular. It's behaviour you can rely on.

---

I don't think text-based predicate comparison is ever going to be a good thing, it's too fragile, even with whitespace normalisation. Non-string lambda functions are superior in every other way so it's natural for a developer to use those instead, especially as sorting predicates become more complex than the basic `"a < b"` form. If the programmer starts out with a simple string lambda then a few months down the line changes it to a non-trivial one - it could even just be a slightly different string lambda that they forgot to change in all relevant locations - it silently eludes the specialization and suddenly the algorithmic complexity of the program is potentially very different, without warning. I think predicate specialization should be robust, and if we can't make it robust, put the onus on the caller instead (which is the status quo).

In lieu of the ability to compare lambda functions, which I don't think anyone has come up with a good design proposal for yet, maybe we could devise a special scheme for sorting predicates in particular, recognizing that in real programs the typical ways they deviate from trivial comparisons is probably a small set. In the thread I linked previously someone suggested predefined predicates `less`, `greater` etc. Maybe something that builds on that, such as with function chaining; `"a.foo < b.foo"` could become `less.member("foo")` or something. Grepping through dub packages could give us some data on what kind of predicates people use in the real world. Just an idea.

---

Regardless of whether or not I'm right about the above, I think you should move the `*Element` algorithms to a new PR where they can be evaluated separately.

---

At any rate, I'm glad you came up with some many algorithms that could benefit from specialization of `SortedRange` inputs. Regardless of what happens these are good finds.

**Comment by JackStouffer** [comment]
@nordlow any chance of moving some of the changes like JakobOvrum suggested? It would be cool to get these into the next release.

**Comment by quickfur** [comment]
I suggest to break this PR up into smaller pieces so that it's easier to review, and the non-controversial parts can be merged first while we work on the other parts. Otherwise this will go very slowly and take too long to get merged.

**Comment by quickfur** [comment]
As a start, I'd say do a separate PR per overload set, as a rough guideline. I think that's much easier to review, and safe to merge piecemeal.

**Comment by quickfur** [comment]
ping @nordlow 
Let's move forward with this?

**Comment by nordlow** [comment]
@quickfur Should I start with an initial pull for the new predicate function `std.range.primitives.isSortedRange` eventhough I have not not found an ideal way to detect which predicate function that was given as template argument to `isSortedRange`?

**Comment by quickfur** [comment]
Hmm. On second thoughts, there is no good way to compare two lambdas (whether string or function literal -- the latter because the compiler has no implementation of such a comparison). So there is no good solution for moving this PR forward.

The problem is that comparing two arbitrary lambdas is, in the most general case, uncomputable. For most practical purposes, though, we can reduce the problem to something tractable by ignoring non-trivial equivalences such as `(a < b) == !(a >= b)`, and just looking at AST equivalence. Of course, that also needs some further restriction, since `a < b+x` may mean different things if `x` is bound to the surrounding context of the lambda, so if they are bound to two different contexts they should not compare as equal.

But in any case, this requires compiler support, and I agree with the sentiment that we should not promote string lambdas anymore so it's kinda pointless to support these functions just for string lambdas.  Let's wait until we have a DIP on how to implement lambda comparisons.

**Comment by MetaLang** [comment]
> Of course, that also needs some further restriction, since a < b+x may mean different things if x is bound to the surrounding context of the lambda, so if they are bound to two different contexts they should not compare as equal.

I'm not sure if you're saying that this is bad, but I'd consider that to be a pretty good thing.

**Comment by quickfur** [comment]
I was just thinking aloud.  It would be very bad if the two lambdas in the following code compared equal!
````
auto sortedRange1(int[] a)
{
    auto x = 1;
    return a.sort!((a,b) => a < x*b);
}
auto sortedRange2(int[] a)
{
    auto x =  -1;
    return a.sort((a,b) => a < x*b);
}
````

**Comment by sigod** [comment]
I have a question: Why not to add specializations directly into `SortedRange`?

**Comment by nordlow** [comment]
That might be cleaner, thanks to UFCS.

AFAIK: Lambda comparison problem will remain, though.

**Comment by quickfur** [comment]
Actually, if the specializations are added as member functions of `SortedRange`, then we don't *need* `isSortedRange` and lambda comparisons, because the member functions will have direct access to the sorting predicate and the range can safely be assumed to be sorted (since otherwise it wouldn't be a `SortedRange` to begin with).

Given the current state of things, that could potentially be the better way to go right now.

**Comment by nordlow** [comment]
Ok, I'll look into it. 

I guess we should be begin with:

- `sort`
- `isSorted`
- `minPos`

Anymore low-hanging fruits?

But note that this, of course, only works for calls to `sort`, `isSorted`, `minPos` without explicit lambda-argument `pred`. But I guess that's ok, right?

**Comment by sigod** [comment]
> then we don't need ... lambda comparisons

There's slight problem. For example, if range were sorted with `"a < b"` and user wants to sort it again with `"a > b"`. Ideally it's just a call to `retro`.

**Comment by sigod** [comment]
> Anymore low-hanging fruits?

`minCount`, probably.

**Comment by quickfur** [comment]
@sigod  We can't possibly know about all these special cases.  It would be up to user code to detect such a case and use `retro` instead of `sort`.  (Of course, ideally, `sort` should be using an algorithm that behaves pretty closely to calling `retro` when the incoming range is sorted in reverse order.) But there's also the question of how common is it to need to call `sort` on a sorted range with the reverse predicate, to justify including this special case in the standard library.  If this is a rare use case, it doesn't justify the cost of additional complexity in `SortedRange`.

(Not to mention that determining whether two predicates are the opposite of each other, generally speaking, is uncomputable. It's easy to detect built-in operators `<` and `>`, but what if the user type implements a non-trivial `opCmp`? What if `opCmp` is a partial order, rather than a total order? Etc.)

**Comment by sigod** [comment]
> (Not to mention that determining whether two predicates are the opposite of each other, generally speaking, is uncomputable. It's easy to detect built-in operators `<` and `>`, but what if the user type implements a non-trivial `opCmp`? What if `opCmp` is a partial order, rather than a total order? Etc.)

This didn't occurred to me. Then I think such detection is out of question.

> But note that this, of course, only works for calls to `sort`, `isSorted`, `minPos` without explicit lambda-argument `pred`. But I guess that's ok, right?

We probably could easily add support for the same `pred` with which `SortedRange` were constructed.

**Comment by wilzbach** [comment]
@nordlow `{min,max}Element` are now in Phobos (crowd goes woohoo!). I actually didn't know you also had them in this PR ;-)
Thus could you please remove them from your PR?

> Actually, if the specializations are added as member functions of SortedRange, then we don't need isSortedRange and lambda comparisons

What else was blocking this PR?

**Comment by nordlow** [comment]
But we still need overloads for {min,max}Element in the case of `SortedRange`, right?

**Comment by wilzbach** [comment]
> I'll restrict the overloads in this pull to only operate on SortedRange with a predicate that matches the predicate of minElement.

1) Please use `static if` - we should put such details to the implementation. Have a look at the post by [big boss](http://forum.dlang.org/thread/nflotk$1ejk$1@digitalmars.com).

2) You should also have a look at #4257 - it's an "optimization" for the default case of an empty lambda.

3) How do you know that the mapping function if given and the sorting pred match? -> imho we can only make the `static if` if no mapping function is used (see 2))

**Comment by sigod** [comment]
> Should these overloads be implemented as free functions take a SortedRange as argument or via SortedRange-members?

I vote for `SortedRange`-members. It's easier and cleaner.

**Comment by wilzbach** [comment]
> I vote for SortedRange-members. It's easier and cleaner.

How do you avoid that SortedRanges match the other template? Won't you have to add something like `!isSortedRange!R`

**Comment by sigod** [comment]
> How do you avoid that SortedRanges match the other template?

It's intended for `SortedRange`-members to "override" some of the common templates.

> Won't you have to add something like !isSortedRange!R

I hope not.

**Comment by nordlow** [comment]
But to support overloading of `minElement(Range, Seed)` don't we need to override behaviour through a *free* function overload

``D
minElement(R)(R r)
    if (isSortedRange!R)
{ ... }
```

?

Or does D's UFCS work bothways? I dont' remember. I only make use of it as: `r.f(...)` => `f(r, ...)`, not vice versa.

**Comment by sigod** [comment]
> `minElement(Range, Seed)`

Yes, for such use we need free function overload. To be honest I didn't consider it. I always use UFCS.

> or does D's UFCS work bothways?

No. Only `f(r, ...) => r.f(...)`.

**Comment by nordlow** [comment]
So, AFAICT, the only current solution to provide specialization is via member-call syntax then, right?

**Comment by sigod** [comment]
> So, AFAICT, the only current solution is to provide specialization is via member-call syntax then, right?

That or we need to deal with comparison of predicates. I'd go with former.

**Comment by JackStouffer** [comment]
Closing due to inactivity. @nordlow If you want to reopen this and continue working on it, please let me know.
